File: string/rotate string.java
class Solution {
    public boolean rotateString(String A, String B) {
        if (A == null || B == null || A.length() != B.length()) {
            return false;
        }
        if (A.equals(B)) {
            return true;
        }
        // move characters, one at a time, and compare
        StringBuffer sb = new StringBuffer(A);
        for (char c : A.toCharArray()) {
            sb.deleteCharAt(0);
            sb.append(c);
            if (sb.toString().equals(B)) {
                return true;
            }
        }
        return false;
    }
}

File: string/rotate string.cpp
class Solution {
public:
    bool rotateString(string A, string B) {
         if(A.size() != B.size()) return false;
        if(A.size() == 0) return true;
        
        vector<int> candidateA;
        
        for(int i = 0; i < A.size(); i++){
            if(A[i] == B[0]){
                candidateA.push_back(i);
            }
        }
        
        for(int start : candidateA){
            bool isRotate = true;
            for(int i = 0; i < A.size(); i++){
                if(A[(start+i)%A.size()] != B[i]){
                    isRotate = false;
                    break;
                }
            }
            if(isRotate){
                return true;
            }
        }
        
        return false;
    }
};
